### PR TITLE
`recruitment/create-candidate` 2.0 profile

### PR DESCRIPTION
**Pull request by janhalama**
## Description

This PR adds `recruitment/create-candidate` 2.0 profile version with following changes:
- `cv` input fields changed from { data: string, name: string } to { data, fileName: string }, type of data is not specified, but it accepts `BinaryData` and `name` field has been changed to `fileName` as it better describes its purpose.
- new use case `CreateCandidateFeatures` added with list of features supported by the provider
- `CVFileNameRequired` and `CVMIMETypeNotSupported` error types added

Based on #324 

## Motivation and Context

Support CV upload for `breezy-hr` provider.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] I haven't repeated the code. (DRY)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: grid/recruitment/create-candidate/profile.supr [hunk 1]
```diff
@@ -139,6 +139,45 @@ usecase CreateCandidate {
...
+      The cv input contains url field with publicly accessible URL of a CV file.
+      """
+      url = 'url'
+      """
+      File
+      The cv input contains data field with CV file and fileName field with the file name.
+      """
+      file = 'file'
+    }!
+  ]!
```
Comment (martinalbert) [inline grid/recruitment/create-candidate/profile.supr:170]:
```suggestion
    """
    CV MIME Types
    List of CV MIME types accepted by the provider. MIME types are defined in RFC2045 and registered by https://www.iana.org/.
    """
    cvMIMETypes! [string!]!
    """
    CV Upload Methods
    List of supported CV upload methods.
    """
    cvUploadMethods! [
      enum {
        """
        URL
        The cv input contains url field with publicly accessible URL of a CV file.
        """
        url = 'url'
        """
        File
        The cv input contains data field with CV file and fileName field with the file name.
        """
        file = 'file'
      }!
    ]!
```

File: grid/recruitment/create-candidate/maps/breezy-hr.suma [hunk 2]
```diff
@@ -242,3 +264,66 @@ operation CheckCandidateByEmail {
...
+operation UploadCV {
+  cv = args.candidate.cv || {}
+
+  return if(cv.data === undefined || cv.data === null) {}
+
+  tmp = (cv.data.name = cv.fileName)
+
+  http POST "/v3/company/{ args.companyId }/position/{ args.candidate.jobId }/candidate/{ args.candidateId }/resume" {
+
+  request "multipart/form-data" {
```
Comment (martinalbert) [inline grid/recruitment/create-candidate/maps/breezy-hr.suma:297]:
```suggestion
    request "multipart/form-data" {
```

File: grid/recruitment/create-candidate/maps/breezy-hr.test.ts [hunk 3]
```diff
@@ -1,6 +1,53 @@
...
+          //resume upload contains multipart/form-data body with boundary with randomized value, we replace the boundary with constant in recorded body and matched body
+          const recordedBody = replaceBoundaryInMultipartFormDataBody(
+            boundaryReplacement,
+            recording.body
+          );
+          recording.body = body => {
+            const requestBody = replaceBoundaryInMultipartFormDataBody(
+              boundaryReplacement,
+              body
+            );
```
Comment (martinalbert) [inline grid/recruitment/create-candidate/maps/breezy-hr.test.ts:31]: question: why is this needed? 🤔

File: grid/recruitment/create-candidate/maps/recordings/breezy-hr.recording.json [hunk 4]
```diff
@@ -255,16 +280,237 @@
...
       }
+    ],
+    "0de0fe70126c707d7921a8307e36c0ec": [],
+    "29a5bf03ceca5dbd3e7d244d42b01174": [],
+    "4d58e4b9dbfa220a6ddad83fe9d25e7e": [],
+    "8d1596557a6ba534defec4d4d9016532": [],
+    "cfa81861f9e39455b6c83fc15d1942da": [],
+    "f4dc24e7230ff97506452f9a9ecaa2c2": [],
+    "1f557bdc0faa8ff1c45d6772a897d773": [
+      {
```
Comment (martinalbert) [inline grid/recruitment/create-candidate/maps/recordings/breezy-hr.recording.json:301]: Aren't these duplicated? (see also bellow)

File: grid/recruitment/create-candidate/maps/recordings/workable.recording.json [hunk 5]
```diff
@@ -249,10 +249,212 @@
...
+        ],
+        "responseIsBinary": false
+      }
+    ],
+    "d92c273026c27383eaa78974b272b2ff": [],
+    "1a1c1f0b9161bd1f6848dbafdeec6e46": [],
+    "f5d3ca6e0319573833951afff46d2008": [],
+    "e14b0dd6837efbd20652e1a90b21ceef": [],
+    "323c2ff09bf66f9d2eb60ab73e0d2095": [
+      {
```
Comment (martinalbert) [inline grid/recruitment/create-candidate/maps/recordings/workable.recording.json:262]: Also seems like duplication.

File: grid/recruitment/create-candidate/maps/breezy-hr.test.ts (hunk 3 above)
Comment (janhalama) [reply grid/recruitment/create-candidate/maps/breezy-hr.test.ts:31]: The boundary needs to be replaced with constant value in recorded and tested requests because boundary changes with each subsequent request even when the rest of the request contains the same data.

This will be the case of every multipart/form-data request tested using nock and @superfaceai/testing lib. We should come with generic solution in @superfaceai/testing library.

File: grid/recruitment/create-candidate/maps/breezy-hr.test.ts [hunk 6]
```diff
@@ -1,6 +1,53 @@
...
+    beforeRecordingLoad: recordings => {
+      recordings.forEach(recording => {
+        if (
+          recording.method === 'POST' &&
+          /^\/v3\/company\/.+\/position\/.+\/candidate\/.+\/resume$/m.test(
+            recording.path
+          )
+        ) {
+          const boundaryReplacement = '---boundary replaced';
+          //resume upload contains multipart/form-data body with boundary with randomized value, we replace the boundary with constant in recorded body and matched body
```
Comment (janhalama) [inline grid/recruitment/create-candidate/maps/breezy-hr.test.ts:22]:
```suggestion
          //Requests with multipart/form-data content type contains boundary in request body. Boundary value changes even when the rest of the request contains the same data.
```

File: grid/recruitment/create-candidate/maps/breezy-hr.test.ts (hunk 3 above)
Comment (janhalama) [reply grid/recruitment/create-candidate/maps/breezy-hr.test.ts:31]: See my suggestion with better description in comment above.

File: grid/recruitment/create-candidate/maps/recordings/workable.recording.json (hunk 5 above)
Comment (janhalama) [reply grid/recruitment/create-candidate/maps/recordings/workable.recording.json]: fixed

File: grid/recruitment/create-candidate/maps/recordings/breezy-hr.recording.json (hunk 4 above)
Comment (janhalama) [reply grid/recruitment/create-candidate/maps/recordings/breezy-hr.recording.json]: Fixed, but I am not sure what is the right testing flow to avoid duplications. Looks like I have to delete file with recording before calling 
```
yarn test:record grid/recruitment/create-candidate/maps/breezy-hr -u
```
in order to avoid duplicate snapshots.

Comment (martinalbert) [reply grid/recruitment/create-candidate/maps/recordings/breezy-hr.recording.json]: Currently the best way is to always try the test with mocked traffic and either:
- fix a hash for an old recording to one that can't be found (Mentioned in error message)
- or remove recordings that you think are duplicated and rerecord

File: grid/recruitment/create-candidate/maps/breezy-hr.test.ts (hunk 3 above)
Comment (martinalbert) [reply grid/recruitment/create-candidate/maps/breezy-hr.test.ts:31]: I understand and agree that this should be added in processing of recordings logic in testing library. I have two more questions:
- why do you assign function to a body?
- isn't there a need for `beforeRecordingSave` implementation? Does it work only with this (record -> test with recorded traffic)?

Comment (janhalama) [reply grid/recruitment/create-candidate/maps/breezy-hr.test.ts:31]: > why do you assign function to a body?

The function is custom body matcher implementation. It is supported by nock library.

> isn't there a need for beforeRecordingSave implementation

Well we could change the body before we save the recording, but that body would still not match the body generated during the test run. Boundary is random value.

Comment (martinalbert) [reply grid/recruitment/create-candidate/maps/breezy-hr.test.ts:31]: Thanks for clarifying, was just wondering if we don't want to replace boundary in `beforeRecordingSave` and keep recording without this dynamic boundary, my only concern is that recording body will change each time we rerecord it and it is not clear what changed in hex format. 🤷🏻‍♂️

**Review by martinalbert — APPROVED** [review]
LGTM

File: grid/recruitment/create-candidate/maps/breezy-hr.test.ts (hunk 3 above)
Comment (janhalama) [reply grid/recruitment/create-candidate/maps/breezy-hr.test.ts:31]: > just wondering if we don't want to replace boundary in beforeRecordingSave and keep recording without this dynamic boundary

It is the request body that is changing, not response body. Keep recording without dynamic boundary would not help, because the request body boundary generated during test run would not match saved recording.

Comment (martinalbert) [reply grid/recruitment/create-candidate/maps/breezy-hr.test.ts:31]: Wouldn't it match recordings with your implementation of `beforeRecordingLoad` (with use of function as a body matcher)?
